File: c/gtk3/app-chooser-widget.c
#include <gtk/gtk.h>

typedef enum {
  SHOW_ALL,
  SHOW_DEFAULT,
  SHOW_FALLBACK,
  SHOW_OTHER,
  SHOW_RECOMMENDED
} AppChooserWidgetSetting;

typedef struct _ToggleButtonData {
  GPtrArray* widgets;
  AppChooserWidgetSetting setting;
} ToggleButtonData;

static void on_app_activate(GApplication* self, gpointer data);
static void on_app_startup(GApplication* self, gpointer data);
static void on_app_shutdown(GApplication* self, gpointer data);
static void on_button_toggled(GtkToggleButton* self, gpointer data);

const static gchar* APP_ID = "io.github.Miqueas.GTK-Examples.C.Gtk3.AppChooserWidget";
const static gchar* APP_TITLE = "GtkAppChooserWidget";

gint main(gint argc, gchar** argv) {
  GtkApplication* app = gtk_application_new(APP_ID, 0);
  g_signal_connect(app, "startup", G_CALLBACK(on_app_startup), NULL);
  g_signal_connect(app, "activate", G_CALLBACK(on_app_activate), NULL);

  gint result = g_application_run(G_APPLICATION(app), argc, argv);
  g_object_unref(app);

  return result;
}

static void on_app_activate(GApplication* self, gpointer data) {
  GtkWindow* window = gtk_application_get_active_window(GTK_APPLICATION(self));
  if (window != NULL) gtk_window_present(window);
}

static void on_app_startup(GApplication* self, gpointer data) {
  GPtrArray* widgets = g_ptr_array_sized_new(6);
  ToggleButtonData* show_all_data = g_new(ToggleButtonData, 1);
  ToggleButtonData* show_default_data = g_new(ToggleButtonData, 1);
  ToggleButtonData* show_fallback_data = g_new(ToggleButtonData, 1);
  ToggleButtonData* show_other_data = g_new(ToggleButtonData, 1);
  ToggleButtonData* show_recommended_data = g_new(ToggleButtonData, 1);
  GPtrArray* free_data = g_ptr_array_sized_new(6);
  GtkWidget* window = gtk_application_window_new(GTK_APPLICATION(self));
  GtkWidget* box = gtk_box_new(GTK_ORIENTATION_VERTICAL, 10);
  GtkWidget* app_chooser_widget = gtk_app_chooser_widget_new("image/png");
  GtkWidget* settings_label = gtk_label_new("Settings:");
  GtkWidget* show_all_toggle = gtk_check_button_new_with_label("Show all applications");
  GtkWidget* show_default_toggle = gtk_check_button_new_with_label("Show default");
  GtkWidget* show_fallback_toggle = gtk_check_button_new_with_label("Show fallback");
  GtkWidget* show_other_toggle = gtk_check_button_new_with_label("Show other");
  GtkWidget* show_recommended_toggle = gtk_check_button_new_with_label("Show recommended");

  g_ptr_array_add(widgets, app_chooser_widget);
  g_ptr_array_add(widgets, show_all_toggle);
  g_ptr_array_add(widgets, show_default_toggle);
  g_ptr_array_add(widgets, show_fallback_toggle);
  g_ptr_array_add(widgets, show_other_toggle);
  g_ptr_array_add(widgets, show_recommended_toggle);

  show_all_data->widgets = widgets;
  show_all_data->setting = SHOW_ALL;
  show_default_data->widgets = widgets;
  show_default_data->setting = SHOW_DEFAULT;
  show_fallback_data->widgets = widgets;
  show_fallback_data->setting = SHOW_FALLBACK;
  show_other_data->widgets = widgets;
  show_other_data->setting = SHOW_OTHER;
  show_recommended_data->widgets = widgets;
  show_recommended_data->setting = SHOW_RECOMMENDED;

  g_ptr_array_add(free_data, widgets);
  g_ptr_array_add(free_data, show_all_data);
  g_ptr_array_add(free_data, show_default_data);
  g_ptr_array_add(free_data, show_fallback_data);
  g_ptr_array_add(free_data, show_other_data);
  g_ptr_array_add(free_data, show_recommended_data);

  gtk_container_add(GTK_CONTAINER(window), box);
  gtk_window_set_title(GTK_WINDOW(window), APP_TITLE);
  gtk_window_set_default_size(GTK_WINDOW(window), 400, 400);
  gtk_container_set_border_width(GTK_CONTAINER(window), 10);

  gtk_box_pack_start(GTK_BOX(box), app_chooser_widget, TRUE, TRUE, 0);
  gtk_box_pack_start(GTK_BOX(box), settings_label, FALSE, FALSE, 0);
  gtk_box_pack_start(GTK_BOX(box), show_all_toggle, FALSE, FALSE, 0);
  gtk_box_pack_start(GTK_BOX(box), show_default_toggle, FALSE, FALSE, 0);
  gtk_box_pack_start(GTK_BOX(box), show_fallback_toggle, FALSE, FALSE, 0);
  gtk_box_pack_start(GTK_BOX(box), show_other_toggle, FALSE, FALSE, 0);
  gtk_box_pack_start(GTK_BOX(box), show_recommended_toggle, FALSE, FALSE, 0);
  gtk_widget_show_all(box);

  gtk_widget_set_halign(settings_label, GTK_ALIGN_START);

  gtk_toggle_button_set_active(
    GTK_TOGGLE_BUTTON(show_all_toggle),
    gtk_app_chooser_widget_get_show_all(
      GTK_APP_CHOOSER_WIDGET(app_chooser_widget)
    )
  );

  gtk_toggle_button_set_active(
    GTK_TOGGLE_BUTTON(show_default_toggle),
    gtk_app_chooser_widget_get_show_default(
      GTK_APP_CHOOSER_WIDGET(app_chooser_widget)
    )
  );

  gtk_toggle_button_set_active(
    GTK_TOGGLE_BUTTON(show_fallback_toggle),
    gtk_app_chooser_widget_get_show_fallback(
      GTK_APP_CHOOSER_WIDGET(app_chooser_widget)
    )
  );

  gtk_toggle_button_set_active(
    GTK_TOGGLE_BUTTON(show_other_toggle),
    gtk_app_chooser_widget_get_show_other(
      GTK_APP_CHOOSER_WIDGET(app_chooser_widget)
    )
  );

  gtk_toggle_button_set_active(
    GTK_TOGGLE_BUTTON(show_recommended_toggle),
    gtk_app_chooser_widget_get_show_recommended(
      GTK_APP_CHOOSER_WIDGET(app_chooser_widget)
    )
  );

  g_signal_connect(
    show_all_toggle,
    "toggled",
    G_CALLBACK(on_button_toggled),
    show_all_data
  );

  g_signal_connect(
    show_default_toggle,
    "toggled",
    G_CALLBACK(on_button_toggled),
    show_default_data
  );

  g_signal_connect(
    show_fallback_toggle,
    "toggled",
    G_CALLBACK(on_button_toggled),
    show_fallback_data
  );

  g_signal_connect(
    show_other_toggle,
    "toggled",
    G_CALLBACK(on_button_toggled),
    show_other_data
  );

  g_signal_connect(
    show_recommended_toggle,
    "toggled",
    G_CALLBACK(on_button_toggled),
    show_recommended_data
  );

  g_signal_connect(
    self,
    "shutdown",
    G_CALLBACK(on_app_shutdown),
    free_data
  );
}

static void on_app_shutdown(GApplication* self, gpointer data) {
  GPtrArray* free_data = (GPtrArray*) data;
  GPtrArray* widgets = (GPtrArray*) free_data->pdata[0];
  ToggleButtonData* show_all_data = (ToggleButtonData*) free_data->pdata[1];
  ToggleButtonData* show_default_data = (ToggleButtonData*) free_data->pdata[2];
  ToggleButtonData* show_fallback_data = (ToggleButtonData*) free_data->pdata[3];
  ToggleButtonData* show_other_data = (ToggleButtonData*) free_data->pdata[4];
  ToggleButtonData* show_recommended_data = (ToggleButtonData*) free_data->pdata[5];

  g_free(show_all_data);
  g_free(show_default_data);
  g_free(show_fallback_data);
  g_free(show_other_data);
  g_free(show_recommended_data);
  g_ptr_array_free(widgets, TRUE);
  g_ptr_array_free(free_data, TRUE);

  free_data = NULL;
  widgets = NULL;
  show_all_data = NULL;
  show_default_data = NULL;
  show_fallback_data = NULL;
  show_other_data = NULL;
  show_recommended_data = NULL;
}

static void on_button_toggled(GtkToggleButton* self, gpointer data) {
  ToggleButtonData* _data = (ToggleButtonData*) data; 
  GPtrArray* widgets = _data->widgets;
  AppChooserWidgetSetting setting = _data->setting;
  GtkAppChooserWidget* app_chooser_widget = GTK_APP_CHOOSER_WIDGET(widgets->pdata[0]);
  GtkToggleButton* show_all_toggle = GTK_TOGGLE_BUTTON(widgets->pdata[1]);
  gboolean show_all = gtk_toggle_button_get_active(show_all_toggle);
  gboolean is_active = gtk_toggle_button_get_active(self);

  switch (setting) {
    case SHOW_ALL:
      gtk_app_chooser_widget_set_show_all(app_chooser_widget, is_active);
      if (is_active) {
        for (guint i = 2; i < widgets->len; i++) {
          GtkToggleButton* toggle = GTK_TOGGLE_BUTTON(widgets->pdata[i]);
          gtk_toggle_button_set_active(toggle, FALSE);
        }
      }
      break;
    case SHOW_DEFAULT:
      gtk_app_chooser_widget_set_show_default(app_chooser_widget, is_active);
      if (show_all && is_active) gtk_toggle_button_set_active(show_all_toggle, FALSE);
      break;
    case SHOW_FALLBACK:
      gtk_app_chooser_widget_set_show_fallback(app_chooser_widget, is_active);
      if (show_all && is_active) gtk_toggle_button_set_active(show_all_toggle, FALSE);
      break;
    case SHOW_OTHER:
      gtk_app_chooser_widget_set_show_other(app_chooser_widget, is_active);
      if (show_all && is_active) gtk_toggle_button_set_active(show_all_toggle, FALSE);
      break;
    case SHOW_RECOMMENDED:
      gtk_app_chooser_widget_set_show_recommended(app_chooser_widget, is_active);
      if (show_all && is_active) gtk_toggle_button_set_active(show_all_toggle, FALSE);
      break;
    default: break;
  }
}

File: zig/gtk3/app-chooser-widget.zig
const std = @import("std");
const gtk = @import("gtk");
const gio = @import("gio");
const gobject = @import("gobject");

const AppChooserWidgetSetting = enum {
    show_all,
    show_other,
    show_default,
    show_fallback,
    show_recommended,
};

const ToggleData = struct {
    widgets: *[6]*gtk.Widget,
    setting: AppChooserWidgetSetting,
};

const ShutdownData = struct {
    widgets: *[6]*gtk.Widget,
    toggles_data: [5]*ToggleData,
};

var allocator = std.heap.DebugAllocator(.{ .never_unmap = true, .retain_metadata = true }).init;
const gpa = allocator.allocator();
const APP_ID = "io.github.Miqueas.GTK-Examples.Zig.Gtk3.AppChooserWidget";
const APP_TITLE = "GtkAppChooserWidget";

pub fn main(init: std.process.Init.Minimal) void {
    const app = gtk.Application.new(APP_ID, .{});
    defer app.unref();

    _ = gio.Application.signals.startup.connect(app, ?*anyopaque, &onStartup, null, .{});
    _ = gio.Application.signals.activate.connect(app, ?*anyopaque, &onActivate, null, .{});

    const argv = init.args.vector;
    const status = gio.Application.run(
        app.as(gio.Application),
        @intCast(argv.len),
        @ptrCast(@constCast(argv.ptr))
    );

    std.process.exit(@intCast(status));
}

fn onActivate(app: *gtk.Application, _: ?*anyopaque) callconv(.c) void {
    if (app.getActiveWindow()) |window| window.present();
}

fn onShutdown(_: *gtk.Application, data: *ShutdownData) callconv(.c) void {
    for (data.toggles_data) |toggle_data| gpa.destroy(toggle_data);
    gpa.destroy(data.widgets);
    gpa.destroy(data);

    const result = allocator.deinit();
    if (result == .leak) @panic("Leak");
}

fn onStartup(self: *gtk.Application, _: ?*anyopaque) callconv(.c) void {
    const widgets = gpa.create([6]*gtk.Widget) catch @panic("OOM");
    const shutdown_data = gpa.create(ShutdownData) catch @panic("OOM");
    const show_all_data = gpa.create(ToggleData) catch @panic("OOM");
    const show_other_data = gpa.create(ToggleData) catch @panic("OOM");
    const show_default_data = gpa.create(ToggleData) catch @panic("OOM");
    const show_fallback_data = gpa.create(ToggleData) catch @panic("OOM");
    const show_recommended_data = gpa.create(ToggleData) catch @panic("OOM");
    const window = gtk.ApplicationWindow.new(self);
    const box = gtk.Box.new(.vertical, 10);
    const app_chooser_widget = gtk.AppChooserWidget.new("image/png");
    const settings_label = gtk.Label.new("Settings:");
    const show_all_toggle = gtk.CheckButton.newWithLabel("Show all applications");
    const show_other_toggle = gtk.CheckButton.newWithLabel("Show other");
    const show_default_toggle = gtk.CheckButton.newWithLabel("Show default");
    const show_fallback_toggle = gtk.CheckButton.newWithLabel("Show fallback");
    const show_recommended_toggle = gtk.CheckButton.newWithLabel("Show recommended");

    widgets.* = .{
        app_chooser_widget.as(gtk.Widget),
        show_all_toggle.as(gtk.Widget),
        show_other_toggle.as(gtk.Widget),
        show_default_toggle.as(gtk.Widget),
        show_fallback_toggle.as(gtk.Widget),
        show_recommended_toggle.as(gtk.Widget),
    };

    shutdown_data.* = .{
        .widgets = widgets,
        .toggles_data = .{
            show_all_data,
            show_other_data,
            show_default_data,
            show_fallback_data,
            show_recommended_data,
        },
    };

    show_all_data.* = .{ .widgets = widgets, .setting = .show_all, };
    show_other_data.* = .{ .widgets = widgets, .setting = .show_other, };
    show_default_data.* = .{ .widgets = widgets, .setting = .show_default, };
    show_fallback_data.* = .{ .widgets = widgets, .setting = .show_fallback, };
    show_recommended_data.* = .{ .widgets = widgets, .setting = .show_recommended, };

    gtk.Window.setTitle(window.as(gtk.Window), APP_TITLE);
    gtk.Window.setDefaultSize(window.as(gtk.Window), 400, 400);
    gtk.Container.add(window.as(gtk.Container), box.as(gtk.Widget));
    gtk.Container.setBorderWidth(window.as(gtk.Container), 10);

    box.packStart(app_chooser_widget.as(gtk.Widget), 1, 1, 0);
    box.packStart(settings_label.as(gtk.Widget), 0, 0, 0);
    box.packStart(show_all_toggle.as(gtk.Widget), 0, 0, 0);
    box.packStart(show_default_toggle.as(gtk.Widget), 0, 0, 0);
    box.packStart(show_fallback_toggle.as(gtk.Widget), 0, 0, 0);
    box.packStart(show_other_toggle.as(gtk.Widget), 0, 0, 0);
    box.packStart(show_recommended_toggle.as(gtk.Widget), 0, 0, 0);
    gtk.Widget.showAll(box.as(gtk.Widget));

    gtk.Widget.setHalign(settings_label.as(gtk.Widget), .start);

    gtk.ToggleButton.setActive(
        show_all_toggle.as(gtk.ToggleButton),
        app_chooser_widget.getShowAll(),
    );

    gtk.ToggleButton.setActive(
        show_default_toggle.as(gtk.ToggleButton),
        app_chooser_widget.getShowDefault(),
    );

    gtk.ToggleButton.setActive(
        show_fallback_toggle.as(gtk.ToggleButton),
        app_chooser_widget.getShowFallback(),
    );

    gtk.ToggleButton.setActive(
        show_other_toggle.as(gtk.ToggleButton),
        app_chooser_widget.getShowOther(),
    );

    gtk.ToggleButton.setActive(
        show_recommended_toggle.as(gtk.ToggleButton),
        app_chooser_widget.getShowRecommended(),
    );

    _ = gio.Application.signals.shutdown.connect(
        self,
        *ShutdownData,
        onShutdown,
        shutdown_data,
        .{},
    );

    _ = gtk.ToggleButton.signals.toggled.connect(
        show_all_toggle,
        *ToggleData,
        onButtonToggled,
        show_all_data,
        .{},
    );

    _ = gtk.ToggleButton.signals.toggled.connect(
        show_other_toggle,
        *ToggleData,
        onButtonToggled,
        show_other_data,
        .{},
    );

    _ = gtk.ToggleButton.signals.toggled.connect(
        show_default_toggle,
        *ToggleData,
        onButtonToggled,
        show_default_data,
        .{},
    );

    _ = gtk.ToggleButton.signals.toggled.connect(
        show_fallback_toggle,
        *ToggleData,
        onButtonToggled,
        show_fallback_data,
        .{},
    );

    _ = gtk.ToggleButton.signals.toggled.connect(
        show_recommended_toggle,
        *ToggleData,
        onButtonToggled,
        show_recommended_data,
        .{},
    );
}

fn onButtonToggled(self: *gtk.CheckButton, data: *ToggleData) callconv(.c) void {
    const app_chooser_widget = gobject.ext.cast(gtk.AppChooserWidget, data.widgets.*[0]).?;
    const show_all_toggle = gobject.ext.cast(gtk.ToggleButton, data.widgets.*[1]).?;
    const is_active = gtk.ToggleButton.getActive(self.as(gtk.ToggleButton));
    const show_all = show_all_toggle.getActive();
    const setting = data.setting;

    switch (setting) {
        .show_all => {
            app_chooser_widget.setShowAll(is_active);
            if (is_active != 0) {
                for (2..6) |index| {
                    const toggle = gobject.ext.cast(gtk.ToggleButton, data.widgets.*[index]).?;
                    toggle.setActive(0);
                }
            }
        },
        .show_other => {
            app_chooser_widget.setShowOther(is_active);
            if (show_all != 0 and is_active != 0) show_all_toggle.setActive(0);
        },
        .show_default => {
            app_chooser_widget.setShowDefault(is_active);
            if (show_all != 0 and is_active != 0) show_all_toggle.setActive(0);
        },
        .show_fallback => {
            app_chooser_widget.setShowFallback(is_active);
            if (show_all != 0 and is_active != 0) show_all_toggle.setActive(0);
        },
        .show_recommended => {
            app_chooser_widget.setShowRecommended(is_active);
            if (show_all != 0 and is_active != 0) show_all_toggle.setActive(0);
        },
    }
}